kafka/src/main/java/org/togglz/kafka/KafkaStateRepository.java

```diff
@@ -0,0 +1,495 @@
+ * can be configured with the <i>initializationTimeout</i>. Moreover, it is a good practice to
+ * implement a health check that monitors the consumer lag.
+ *
+ * <p>The underlying Kafka consumer polls for data. The polling interval can be configured
+ * with the <i>pollingTimeout</i>. If an error occurs during the processing of records
+ * fetched from Kafka, the Kafka consumer will be shutdown. Afterwards, {@link #isRunning()}
+ * will return {@code false} and {@link #getFeatureState(Feature)} will return {@code null}.
+ *
+ * <p>To create a new {@link KafkaStateRepository} instance, it is required to define an
+ * inbound topic and an outbound topic. The underlying Kafka consumer fetches data from the
+ * inbound topic and the Kafka producer sends data to the outbound topic. In most cases, the
+ * inbound topic and the outbound topic will be identical. But depending on the cross datacenter
+ * replication strategy, it might be necessary to use different topics and to replicate the data
+ * between them. The inbound topic should be configured to use the cleanup policy <i>compact</i>
+ * since, on the hand, it is used as persistent storage and, on the other, it has to be entirely
+ * consumed during the initialization of the {@link KafkaStateRepository}.
+ *
+ * @author Florian Stefan
+ */
+public class KafkaStateRepository implements AutoCloseable, StateRepository {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KafkaStateRepository.class);
+
+  private final Map<String, FeatureStateStorageWrapper> featureStates;
+  private final FeatureStateConsumer featureStateConsumer;
+  private final FeatureStateProducer featureStateProducer;
+
+  private KafkaStateRepository(Builder builder) {
+    featureStates = new ConcurrentHashMap<>();
+    featureStateConsumer = new FeatureStateConsumer(builder, featureStates::put);
+    featureStateProducer = new FeatureStateProducer(builder);
+
+    featureStateConsumer.start();
+  }
+
+  /**
+   * Returns a new instance of a {@link KafkaStateRepository} builder.
+   *
+   * @return the new {@link KafkaStateRepository} builder
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  @Override
+  public void close() {
+    featureStateConsumer.close();
+    featureStateProducer.close();
+  }
+
+  @Override
+  public FeatureState getFeatureState(Feature feature) {
+    if (isRunning()) {
+      FeatureStateStorageWrapper storageWrapper = featureStates.get(feature.name());
+
+      if (storageWrapper != null) {
+        return featureStateForWrapper(feature, storageWrapper);
+      } else {
+        LOG.warn("Could not find featureState for given feature - fallback to default value.");
+        return null;
+      }
+    } else {
+      LOG.warn("FeatureStateConsumer not running anymore - fallback to default value.");
+      return null;
+    }
+  }
+
+  @Override
+  public void setFeatureState(FeatureState featureState) {
```

Comment:
The defined contract from the API might not be fulfilled: "The repository implementation must ensure that subsequent calls to {@link #getFeatureState(Feature)} return the same state as persisted using this method."

A subsequent call might return the old value until consumer catchup.

Reply:
Good point! I didn't notice that.

Reply:
Done.